src/tiledb/cloud/vcf/vcf_toolbox/transform.py

```diff
@@ -0,0 +1,74 @@
+import functools
+from typing import Callable, TypeVar
+
+import pyarrow as pa
+
+
+def _df_transform_result(table: pa.Table, fn) -> pa.Table:
+    """
+    Transform the pyarrow table using a user-defined function that operates on
+    a pandas dataframe.
+
+    :param table: pyarrow table
+    :param fn: user-defined function to transform the pandas dataframe
+    :return: pyarrow table
+    """
+
+    # Return if the table is empty
+    if table.num_rows == 0:
+        return table
+
+    # Store the original schema
+    original_schema = table.schema
+
+    # Convert arrow table to pandas
+    df = table.to_pandas()
+
+    # Run user-defined transform function
+    df = fn(df)
+
+    # Convert the DataFrame back to a table
+    result = pa.Table.from_pandas(df)
+
+    # Cast the columns to the original data types because pandas may change the
+    # column type for partitions where the results are all null.
+    for field in original_schema:
+        if field.name in result.column_names:
+            result = result.set_column(
+                result.column_names.index(field.name),
+                field.name,
+                pa.array(result.column(field.name).to_pandas(), type=field.type),
+            )
+
+    return result
+
+
+_CT = TypeVar("_CT", bound=Callable)
+
+
+def df_transform(fn: _CT) -> _CT:
+    """
+    A function decorator that allows users to create a user-defined function to
+    transform a pandas dataframe. The decorated function can be passed directly
+    to the `transform_result` parameter of the `tiledb.cloud.vcf.query` function.
+
+    Parameters
+    ----------
+    fn
+        The user-defined function that takes a pandas dataframe as input and
+        returns a pandas dataframe.
+
+    Returns
+    -------
+        A function than can be passed to the `transform_result` parameter of the
+        `tiledb.cloud.vcf.query` function.
+    """
+
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        return functools.partial(
+            _df_transform_result,
+            fn=functools.partial(fn, *args, **kwargs),
+        )
+
+    return wrapper
```

Comment:
I had a whole comment drafted about the typing for this and then it got eaten when I changed the diff view, so in abbreviated form:

- The typing on the `df_transform` decorator isn’t quite right; it implies that the signature of the returned function is the same as the one that is passed in—rather than a function you pass a dataframe in and get a dataframe out, you pass in a
- The use of `functools.partial` here has the potential to cause problems, most notably because if you pass in any positional arguments, then the dataframe won’t be the first argument to `fn`.

So to specify the types of this entire thing much more precisely:

```python
def _df_transform_result(
    table: pa.Table, fn: Callable[[pd.DataFrame], pd.DataFrame]
) -> pa.Table:
    ...


_PS = ParamSpec("_PS")


def df_transform(
    fn: Callable[Concatenate[pd.DataFrame, _PS], pd.DataFrame]
) -> Callable[_PS, Callable[[pa.Table], pa.Table]]:

    def wrapper(*args: _PS.args, **kwargs: _PS.kwargs) -> Callable[[pa.Table], pa.Table]:
        # Editor's note: within this function, the type checker can infer that
        # tbl is a pa.Table and df is a pd.DataFrame.
        return lambda tbl: _df_transform_result(
            tbl,
            lambda df: fn(df, *args, **kwargs),
        )

    return wrapper
```

and the results:

```python
def do_stuff(df: pd.DataFrame, one: int, two: str, *, three: bytes) -> pd.DataFrame:
    raise NotImplementedError()


reveal_type(do_stuff)
# Type of "do_stuff" is "(df: DataFrame, one: int, two: str, *, three: bytes) -> DataFrame"

do_stuff_transform = df_transform(do_stuff)
reveal_type(do_stuff_transform)
# Type of "do_stuff_transform" is "(one: int, two: str, *, three: bytes) -> ((Table) -> Table)"

stuff_doer = do_stuff_transform(1, "two", three=b"three")
reveal_type(stuff_doer)
# Type of "stuff_doer" is "(Table) -> Table"

stuff_done = stuff_doer(pa.Table())
reveal_type(stuff_done)
# Type of "stuff_done" is "Table"
```

Reply:
Thanks @thetorpedodog for the detailed review of this code and catching the issue with `functools`.

It looks like the `ParamSpec` and `Concatenate` type hints are not available in python 3.9. I'll look into alternatives, please let me know if you have any "hints".

Reply:
Never mind, I found those hints in `typing_extensions`.